### PR TITLE
feat: handle logical assignment in no-self-assign

### PR DESCRIPTION
**Pull request by Zzzen**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What rule do you want to change?**
no-self-assign

**Does this change cause the rule to produce more or fewer warnings?**
more warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**

```js
a &= a;  
a |= a;
a &&= a;
a ||= a;
a ??= a;
```

**What does the rule currently do for this code?**
no errors are reported

**What will the rule do after it's changed?**
all lines will be reported.

#### What changes did you make? (Give an overview)
currently, only `=` is handled in `no-self-assign`. But other logical assignment operators work roughly the same as `=`.

`a &&= a` is `a && (a = a)`, which is essentially a noop.

### CODE REVIEW

**Comment by mdjermanovic** [comment]
Hi @Zzzen, thanks for the PR!

We listed adding support for `&&=`, `||=`, and `??=` operators to the `no-self-assign` rule as _maybe_ in https://github.com/eslint/eslint/issues/13569. We're not yet sure if that change makes sense, so I marked this as "evaluating".

As for `&=` and `|=`, I think that's different as it involves a calculation before the assignment, and it doesn't have to be a noop (should be for integers, but e.g, `1.5 | 1.5` is `1`, `null | null` is `0`, etc.), so those checks probably don't belong to `no-self-assign`.

**Comment by Zzzen** [comment]
Ooops, did not notice the issue. IMO, even though `|=` may have side effects on non-integers, `a |= a` is likely a typo and should be disallowed like how object properties are handled.

**Comment by mdjermanovic** [comment]
@eslint/eslint-tsc thoughts about this proposal for the `no-self-assign` rule?

* Should it report `a &&= a`, `a ||= a`, and `a ??= a`? I think this makes sense since these are essentially the same as `a = a`.
* Should it report `a &= a`, and `a |= a`? I think this doesn't belong to this rule, for the reasons described in [this comment](https://github.com/eslint/eslint/pull/14152#issuecomment-788062375).

**Comment by nzakas** [comment]
@mdjermanovic i agree with your reasoning.

**Comment by btmills** [comment]
@mdjermanovic I agree as well.

File: lib/rules/no-self-assign.js [hunk 1]
```diff
@@ -174,7 +174,7 @@ module.exports = {
 
         return {
             AssignmentExpression(node) {
-                if (node.operator === "=") {
+                if (["=", "&=", "|=", "&&=", "||=", "??="].includes(node.operator)) {
```
Comment (mdjermanovic) [inline lib/rules/no-self-assign.js:177]:
```suggestion
                if (["=", "&&=", "||=", "??="].includes(node.operator)) {
```

File: tests/lib/rules/no-self-assign.js [hunk 2]
```diff
@@ -147,6 +147,31 @@ ruleTester.run("no-self-assign", rule, {
...
+
+        // logical assignment
+        {
+            code: "a &= a",
+            errors: [{ messageId: "selfAssignment", data: { name: "a" } }]
+        },
+        {
+            code: "a |= a",
+            errors: [{ messageId: "selfAssignment", data: { name: "a" } }]
+        },
```
Comment (mdjermanovic) [inline tests/lib/rules/no-self-assign.js:160]:
```suggestion
```

It would be nice to also add `"a &= a"` and `"a |= a"` to `valid` test cases.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
@Zzzen sorry for the delay, the change for `&&=`, `||=`, and `??=` is accepted now.

Can you also update the documentation for this rule?

**Comment by Zzzen** [comment]
👌  updated

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**